drivers/sdhc/rcar_mmc.c

```diff
@@ -0,0 +1,2252 @@
+	case SDHC_BUS_WIDTH4BIT:
+		if (data->props.host_caps.bus_4_bit_support) {
+			reg_width = RCAR_MMC_OPTION_WIDTH_4;
+		} else {
+			LOG_ERR("SDHC I/O: 4-bits bus width isn't supported");
+			return -ENOTSUP;
+		}
+		break;
+	case SDHC_BUS_WIDTH8BIT:
+		if (data->props.host_caps.bus_8_bit_support) {
+			reg_width = RCAR_MMC_OPTION_WIDTH_8;
+		} else {
+			LOG_ERR("SDHC I/O: 8-bits bus width isn't supported");
+			return -ENOTSUP;
+		}
+		break;
+	default:
+		return -ENOTSUP;
+	}
+
+	/*
+	 * Do not change the values of these bits
+	 * when the CBSY bit in SD_INFO2 is 1
+	 */
+	ret = rcar_mmc_poll_reg_flags_check_err(dev, RCAR_MMC_INFO2, RCAR_MMC_INFO2_CBSY, 0, false,
+						false, MMC_POLL_FLAGS_TIMEOUT_US);
+	if (ret) {
+		return -ETIMEDOUT;
+	}
+
+	mmc_option_reg = rcar_mmc_read_reg32(dev, RCAR_MMC_OPTION);
+	mmc_option_reg &= ~RCAR_MMC_OPTION_WIDTH_MASK;
+	mmc_option_reg |= reg_width;
+	rcar_mmc_write_reg32(dev, RCAR_MMC_OPTION, mmc_option_reg);
+
+	host_io->bus_width = ios->bus_width;
+
+	LOG_DBG("%s: set bus-width to %d", dev->name, host_io->bus_width);
+	return 0;
+}
+
+/**
+ * set DDR mode on MMC controller according to value inside
+ * ddr_mode field from @ref mmc_rcar_data structure.
+ */
+static int rcar_mmc_set_ddr_mode(const struct device *dev)
+{
+	int ret = 0;
+	uint32_t if_mode_reg;
+	struct mmc_rcar_data *data = dev->data;
+
+	/*
+	 * Do not change the values of these bits
+	 * when the CBSY bit in SD_INFO2 is 1
+	 */
+	ret = rcar_mmc_poll_reg_flags_check_err(dev, RCAR_MMC_INFO2, RCAR_MMC_INFO2_CBSY, 0, false,
+						false, MMC_POLL_FLAGS_TIMEOUT_US);
+	if (ret) {
+		return -ETIMEDOUT;
+	}
+
+	if_mode_reg = rcar_mmc_read_reg32(dev, RCAR_MMC_IF_MODE);
+	if (data->ddr_mode) {
+		/* HS400 mode (DDR mode) */
+		if_mode_reg |= RCAR_MMC_IF_MODE_DDR;
+	} else {
+		/* Normal mode (default, high speed, or SDR) */
+		if_mode_reg &= ~RCAR_MMC_IF_MODE_DDR;
+	}
+	rcar_mmc_write_reg32(dev, RCAR_MMC_IF_MODE, if_mode_reg);
+
+	return 0;
+}
+
+/**
+ * @brief set timing property of MMC
+ *
+ * For now function only can enable DDR mode and call the function for
+ * changing voltage. It is expectable that we change clock using another
+ * I/O option.
+ * @note In/out parameters should be checked by a caller function.
+ *
+ * @param dev MMC device
+ * @param io I/O properties
+ *
+ * @retval 0 I/O was configured correctly
+ * @retval -ENOTSUP: controller does not support these I/O settings
+ * @retval -ETIMEDOUT: card busy flag is set during long time
+ */
+static int rcar_mmc_set_timings(const struct device *dev, struct sdhc_io *ios)
+{
+	int ret;
+	struct mmc_rcar_data *data = dev->data;
+	struct sdhc_io *host_io = &data->host_io;
+	enum sd_voltage new_voltage = host_io->signal_voltage;
+
+	if (host_io->timing == ios->timing) {
+		return 0;
+	}
+
+	if (!host_io->timing) {
+		return 0;
+	}
+
+	data->ddr_mode = 0;
+
+	switch (ios->timing) {
+	case SDHC_TIMING_LEGACY:
+		break;
+	case SDHC_TIMING_HS:
+		if (!data->props.host_caps.high_spd_support) {
```

Comment:
These checks should not be needed. The SD stack will validate host support using `host_caps` fields before attempting to configure a given timing speed

Reply:
It isn't true if we're talking about the MMC stack, and this driver can be used for both SD and MMC cards. Look at the next call chain: `mmc_set_timing` → `mmc_set_hs_timing` (under the condition `ext->device_type.MMC_HS_52_DV`). The setting of the `MMC_HS_52_DV` field relies only on properties of the MMC card. For example, an MMC card can support this mode, but our controller might not support it.